File: src/valuesteward/steward_state.py
"""Unified system state for Value Steward with cross-process safety."""

from __future__ import annotations

import json
import os
import time
from contextlib import contextmanager
from datetime import date as date_cls
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable
from zoneinfo import ZoneInfo

STATE_PATH = Path("data/steward-state.json")
STATE_LOCK_PATH = Path(f"{STATE_PATH}.lock")
LOCK_TIMEOUT_SEC = float(os.getenv("VS_STATE_LOCK_TIMEOUT_MS", "5000")) / 1000.0
LOCK_STALE_SEC = float(os.getenv("VS_STATE_LOCK_STALE_MS", "15000")) / 1000.0
LOCK_RETRY_SEC = float(os.getenv("VS_STATE_LOCK_RETRY_MS", "25")) / 1000.0


def _default_state() -> dict[str, Any]:
    return {
        "current_mode": "INACTIVE",
        "last_run_at": None,
        "last_mode_transition_reason": "initial_boot",
        "last_known_positions": [],
        "trading_enabled": True,
        "force_no_trade": False,
        "control_reason": None,
        "control_updated_at": None,
        "daily_starting_equity": None,
        "last_equity_reset_date": None,
        "executions_today": 0,
        "last_executed_date": None,
        "last_executed_at": None,
        "last_health_email_at": None,
        "last_health_email_date": None,
        "phase1_start_date": None,
        "phase1_milestones_sent": [],
        "phase1_ready_notified": False,
        "last_eod_email_date": None,
        "version": 1,
    }


def _normalize_state(data: dict[str, Any] | None = None) -> dict[str, Any]:
    source = data if isinstance(data, dict) else {}
    state = {**_default_state(), **source}
    milestones = state.get("phase1_milestones_sent") or []
    if isinstance(milestones, list):
        cleaned = sorted(
            {
                int(value)
                for value in milestones
                if isinstance(value, (int, float)) and value > 0
            }
        )
        state["phase1_milestones_sent"] = cleaned
    else:
        state["phase1_milestones_sent"] = []
    state["phase1_ready_notified"] = bool(state.get("phase1_ready_notified"))
    if state.get("phase1_start_date") is not None and not isinstance(
        state.get("phase1_start_date"), str
    ):
        state["phase1_start_date"] = None
    return state


def _market_timezone() -> ZoneInfo:
    tz = os.getenv("VS_MARKET_TIMEZONE") or "America/New_York"
    try:
        return ZoneInfo(tz)
    except Exception:
        return ZoneInfo("America/New_York")


def _coerce_exchange_date(value: Any) -> date_cls | None:
    if value is None:
        return None
    if isinstance(value, date_cls) and not isinstance(value, datetime):
        return value
    if isinstance(value, datetime):
        dt = value if value.tzinfo else value.replace(tzinfo=timezone.utc)
        return dt.astimezone(_market_timezone()).date()
    if isinstance(value, str):
        raw = value.strip()
        if not raw:
            return None
        if len(raw) == 10:
            try:
                return date_cls.fromisoformat(raw)
            except ValueError:
                return None
        try:
            dt = datetime.fromisoformat(raw.replace("Z", "+00:00"))
        except ValueError:
            return None
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.astimezone(_market_timezone()).date()
    return None


def _read_state_unlocked() -> dict[str, Any]:
    if not STATE_PATH.exists():
        return {}
    return json.loads(STATE_PATH.read_text(encoding="utf-8"))


def _write_state_unlocked(state: dict[str, Any]) -> None:
    payload = _normalize_state(state)
    payload["updated_at"] = datetime.now(timezone.utc).isoformat()
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    tmp_path = STATE_PATH.with_name(
        f"{STATE_PATH.name}.{os.getpid()}.{time.time_ns()}.tmp"
    )
    tmp_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    os.replace(tmp_path, STATE_PATH)


def _is_pid_alive(pid: int) -> bool:
    """Best-effort liveness check for a lock-owner PID (POSIX).

    Guards ``pid <= 0`` because ``os.kill(0, 0)`` / negatives signal whole
    process groups, which would falsely read as 'alive'.
    """
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True  # Exists but owned by another user — still alive.
    except OSError:
        return False


@contextmanager
def _state_lock():
    deadline = time.monotonic() + LOCK_TIMEOUT_SEC
    pid_file = STATE_LOCK_PATH / "owner.pid"
    while True:
        try:
            STATE_LOCK_PATH.mkdir()
            pid_file.write_text(str(os.getpid()), encoding="utf-8")
            break
        except FileExistsError:
            try:
                age = time.time() - STATE_LOCK_PATH.stat().st_mtime
                if age > LOCK_STALE_SEC:
                    # Only evict a stale-looking lock if its owner is gone —
                    # never steal it from a live process (the TOCTOU race a
                    # plain age check is vulnerable to). Missing/corrupt PID is
                    # treated as stale (evictable).
                    try:
                        owner_pid = int(pid_file.read_text(encoding="utf-8").strip())
                        owner_alive = _is_pid_alive(owner_pid)
                    except (FileNotFoundError, ValueError, OSError):
                        owner_alive = False
                    if not owner_alive:
                        try:
                            pid_file.unlink(missing_ok=True)
                        except OSError:
                            pass
                        try:
                            STATE_LOCK_PATH.rmdir()
                        except OSError:
                            pass
                        continue
            except OSError:
                pass
            if time.monotonic() >= deadline:
                raise TimeoutError(f"Timed out acquiring state lock for {STATE_PATH}")
            time.sleep(LOCK_RETRY_SEC)
    try:
        yield
    finally:
        try:
            pid_file.unlink(missing_ok=True)
        except OSError:
            pass
        try:
            STATE_LOCK_PATH.rmdir()
        except OSError:
            pass


def load_steward_state() -> dict[str, Any]:
    """Load state with retries to tolerate concurrent atomic renames."""
    for i in range(3):
        try:
            return _normalize_state(_read_state_unlocked())
        except (json.JSONDecodeError, PermissionError, FileNotFoundError):
            if i == 2:
                break
            time.sleep(0.05 * (i + 1))
    return _normalize_state({})


def save_steward_state(state: dict[str, Any]) -> None:
    """Save a complete state payload under a cross-process lock."""
    with _state_lock():
        _write_state_unlocked(state)


def update_steward_state(
    mutator: Callable[[dict[str, Any]], dict[str, Any] | None]
) -> dict[str, Any]:
    """Read-modify-write state under a cross-process lock."""
    with _state_lock():
        current = _normalize_state(_read_state_unlocked())
        draft = dict(current)
        updated = mutator(draft)
        next_state = draft if updated is None else updated
        payload = _normalize_state(next_state)
        _write_state_unlocked(payload)
        return payload


def get_phase1_start_date(state: dict[str, Any] | None = None) -> date_cls | None:
    current = _normalize_state(state) if state is not None else load_steward_state()
    return _coerce_exchange_date(
        current.get("phase1_start_date") or os.getenv("VS_PHASE1_START_DATE")
    )


def is_on_or_after_phase1_start(
    value: Any, state: dict[str, Any] | None = None
) -> bool:
    candidate = _coerce_exchange_date(value)
    if candidate is None:
        return False
    start = get_phase1_start_date(state)
    return start is None or candidate >= start

File: core/stewardState.js
// core/stewardState.js
import fs from "fs/promises";
import {
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  rmdirSync,
  statSync,
  writeFileSync,
} from "fs";
import path from "path";
import { getExchangeDateString } from "./timeUtils.js";

const STATE_PATH = path.join(process.cwd(), "data", "steward-state.json");
const STATE_LOCK_PATH = `${STATE_PATH}.lock`;
const LEGACY_AGENT_STATE_PATH = path.join(process.cwd(), "data", "agent-state.json");
const LOCK_TIMEOUT_MS = Number(process.env.VS_STATE_LOCK_TIMEOUT_MS ?? 5000);
const LOCK_STALE_MS = Number(process.env.VS_STATE_LOCK_STALE_MS ?? 15000);
const LOCK_RETRY_MS = Number(process.env.VS_STATE_LOCK_RETRY_MS ?? 25);

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function sleepSync(ms) {
  const deadline = Date.now() + ms;
  while (Date.now() < deadline) {
    // Busy wait for short sync lock retries only.
  }
}

function defaultState() {
  return {
    current_mode: "INACTIVE",
    last_run_at: null,
    last_mode_transition_reason: "initial_boot",
    last_known_positions: [],
    trading_enabled: true,
    force_no_trade: false,
    control_reason: null,
    control_updated_at: null,
    daily_starting_equity: null,
    last_equity_reset_date: null,
    executions_today: 0,
    last_executed_date: null,
    last_executed_at: null,
    last_health_email_at: null,
    last_health_email_date: null,
    phase1_start_date: null,
    phase1_milestones_sent: [],
    phase1_ready_notified: false,
    last_eod_email_date: null,
    version: 1,
    updated_at: new Date().toISOString(),
  };
}

function loadLegacyAgentStateSync() {
  if (!existsSync(LEGACY_AGENT_STATE_PATH)) return null;
  try {
    return JSON.parse(readFileSync(LEGACY_AGENT_STATE_PATH, "utf8"));
  } catch {
    return null;
  }
}

function normalizeState(data = {}) {
  const source = data && typeof data === "object" ? data : {};
  const next = { ...defaultState(), ...source };
  const legacy = loadLegacyAgentStateSync();

  if (legacy) {
    if (!("last_health_email_at" in source) && legacy.last_health_email_at) {
      next.last_health_email_at = legacy.last_health_email_at;
    }
    if (!("last_health_email_date" in source) && legacy.last_health_email_date) {
      next.last_health_email_date = legacy.last_health_email_date;
    }
    if (!("last_eod_email_date" in source) && legacy.last_eod_email_date) {
      next.last_eod_email_date = legacy.last_eod_email_date;
    }
    if (
      !("phase1_milestones_sent" in source) &&
      Array.isArray(legacy.phase1_milestones_sent)
    ) {
      next.phase1_milestones_sent = legacy.phase1_milestones_sent;
    }
    if (!("phase1_ready_notified" in source) && legacy.phase1_ready_notified === true) {
      next.phase1_ready_notified = true;
    }
  }

  if (next.phase1_start_date !== null && typeof next.phase1_start_date !== "string") {
    next.phase1_start_date = null;
  }
  next.phase1_milestones_sent = Array.isArray(next.phase1_milestones_sent)
    ? Array.from(
        new Set(
          next.phase1_milestones_sent.filter(
            (value) => Number.isFinite(value) && value > 0
          )
        )
      ).sort((a, b) => a - b)
    : [];
  next.phase1_ready_notified = next.phase1_ready_notified === true;

  return next;
}

function stateTempPath() {
  return `${STATE_PATH}.${process.pid}.${Date.now()}.${Math.random()
    .toString(16)
    .slice(2)}.tmp`;
}

function readStateFileSync() {
  if (!existsSync(STATE_PATH)) return {};
  return JSON.parse(readFileSync(STATE_PATH, "utf8"));
}

async function readStateFile() {
  const raw = await fs.readFile(STATE_PATH, "utf8");
  return JSON.parse(raw);
}

function writeStateFileSync(payload) {
  const dir = path.dirname(STATE_PATH);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  const tmpPath = stateTempPath();
  writeFileSync(tmpPath, JSON.stringify(payload, null, 2));
  renameSync(tmpPath, STATE_PATH);
}

async function writeStateFile(payload) {
  const dir = path.dirname(STATE_PATH);
  await fs.mkdir(dir, { recursive: true });
  const tmpPath = stateTempPath();
  await fs.writeFile(tmpPath, JSON.stringify(payload, null, 2));
  await fs.rename(tmpPath, STATE_PATH);
}

async function acquireLock() {
  const deadline = Date.now() + LOCK_TIMEOUT_MS;
  while (Date.now() <= deadline) {
    try {
      await fs.mkdir(STATE_LOCK_PATH);
      return;
    } catch (err) {
      if (err?.code !== "EEXIST") throw err;
      try {
        const stats = await fs.stat(STATE_LOCK_PATH);
        if (Date.now() - stats.mtimeMs > LOCK_STALE_MS) {
          await fs.rmdir(STATE_LOCK_PATH);
          continue;
        }
      } catch {
        // Another process may have released the lock already.
      }
      await sleep(LOCK_RETRY_MS);
    }
  }
  throw new Error(`Timed out acquiring state lock for ${STATE_PATH}`);
}

function acquireLockSync() {
  const deadline = Date.now() + LOCK_TIMEOUT_MS;
  while (Date.now() <= deadline) {
    try {
      mkdirSync(STATE_LOCK_PATH);
      return;
    } catch (err) {
      if (err?.code !== "EEXIST") throw err;
      try {
        const stats = statSync(STATE_LOCK_PATH);
        if (Date.now() - stats.mtimeMs > LOCK_STALE_MS) {
          rmdirSync(STATE_LOCK_PATH);
          continue;
        }
      } catch {
        // Another process may have released the lock already.
      }
      sleepSync(LOCK_RETRY_MS);
    }
  }
  throw new Error(`Timed out acquiring state lock for ${STATE_PATH}`);
}

async function releaseLock() {
  try {
    await fs.rmdir(STATE_LOCK_PATH);
  } catch {
    // Lock may already be gone after a stale-lock cleanup.
  }
}

function releaseLockSync() {
  try {
    rmdirSync(STATE_LOCK_PATH);
  } catch {
    // Lock may already be gone after a stale-lock cleanup.
  }
}

async function withStateLock(fn) {
  await acquireLock();
  try {
    return await fn();
  } finally {
    await releaseLock();
  }
}

function withStateLockSync(fn) {
  acquireLockSync();
  try {
    return fn();
  } finally {
    releaseLockSync();
  }
}

/**
 * Professional Reading: Retry logic to handle concurrent atomic renames.
 */
async function readWithRetry(retries = 3) {
  for (let i = 0; i < retries; i += 1) {
    try {
      const raw = await fs.readFile(STATE_PATH, "utf8");
      return JSON.parse(raw);
    } catch (err) {
      if (i === retries - 1) throw err;
      await sleep(50 * (i + 1));
    }
  }
  return {};
}

export async function loadState() {
  try {
    const data = await readWithRetry();
    return normalizeState(data);
  } catch {
    return normalizeState({});
  }
}

export function loadStateSync() {
  if (!existsSync(STATE_PATH)) return normalizeState({});
  for (let i = 0; i < 3; i += 1) {
    try {
      return normalizeState(readStateFileSync());
    } catch {
      if (i === 2) return normalizeState({});
      sleepSync(25 * (i + 1));
    }
  }
  return normalizeState({});
}

/**
 * Writes a complete state payload under lock.
 */
export async function saveState(state) {
  return withStateLock(async () => {
    const payload = { ...normalizeState(state), updated_at: new Date().toISOString() };
    await writeStateFile(payload);
    return payload;
  });
}

export function saveStateSync(state) {
  return withStateLockSync(() => {
    const payload = { ...normalizeState(state), updated_at: new Date().toISOString() };
    writeStateFileSync(payload);
    return payload;
  });
}

export async function updateState(mutator) {
  return withStateLock(async () => {
    const current = existsSync(STATE_PATH) ? normalizeState(await readStateFile()) : normalizeState({});
    const draft = structuredClone(current);
    const nextState = (await mutator(draft)) ?? draft;
    const payload = { ...normalizeState(nextState), updated_at: new Date().toISOString() };
    await writeStateFile(payload);
    return payload;
  });
}

export function updateStateSync(mutator) {
  return withStateLockSync(() => {
    const current = existsSync(STATE_PATH) ? normalizeState(readStateFileSync()) : normalizeState({});
    const draft = structuredClone(current);
    const nextState = mutator(draft) ?? draft;
    const payload = { ...normalizeState(nextState), updated_at: new Date().toISOString() };
    writeStateFileSync(payload);
    return payload;
  });
}

export async function markHealthEmailSent(sentAt = new Date()) {
  const ts = sentAt instanceof Date ? sentAt : new Date(sentAt);
  return updateState((state) => {
    state.last_health_email_at = ts.toISOString();
    state.last_health_email_date = getExchangeDateString(ts);
    return state;
  });
}

export async function markPhaseEmailSent({ milestones = [], ready = false } = {}) {
  return updateState((state) => {
    const sent = new Set(state.phase1_milestones_sent ?? []);
    for (const milestone of milestones) {
      if (Number.isFinite(milestone) && milestone > 0) {
        sent.add(milestone);
      }
    }
    state.phase1_milestones_sent = Array.from(sent).sort((a, b) => a - b);
    if (ready) {
      state.phase1_ready_notified = true;
    }
    return state;
  });
}

export async function markEodEmailSent(sentAt = new Date()) {
  const ts = sentAt instanceof Date ? sentAt : new Date(sentAt);
  return updateState((state) => {
    state.last_eod_email_date = getExchangeDateString(ts);
    return state;
  });
}

export function markEodEmailSentSync(sentAt = new Date()) {
  const ts = sentAt instanceof Date ? sentAt : new Date(sentAt);
  return updateStateSync((state) => {
    state.last_eod_email_date = getExchangeDateString(ts);
    return state;
  });
}
